Tests failed when running under root permission

https://github.com/twpayne/go-vfs/blob/906833b4a68038e248cb6a5637266557562dc42b/vfst/fs_test.go#L74-L103

`fileSystem.Chmod("/dir/subdir", 0)` will not prevent folder be accessed under root permission

```
github.com/twpayne/go-vfs/vfst                                                                                                                                                                                                                                                
--- FAIL: TestWalkErrors (0.00s)                                                                                                                                                                                                                                              
    --- FAIL: TestWalkErrors/private_subdir (0.00s)                                                                                                                                                                                                                           
        fs_test.go:127: Expected values to be equal:                                                                                                                                                                                                                          
               "/",                                                                                                                                                                                                                                                           
               "/dir",                                                                                                                                                                                                                                                        
               "/dir/subdir",                                                                                                                                                                                                                                                 
            +  "/dir/subdir/subsubdir",                                                                                                                                                                                                                                       
            +  "/dir/subdir/subsubdir/file",                                                                                                                                                                                                                                  
             }                                                                                                                                                                                                                                                                
    --- FAIL: TestWalkErrors/private_subdir_keep_going (0.00s)                                                                                                                                                                                                                
        fs_test.go:127: Expected values to be equal:                                                                                                                                                                                                                          
               "/",                                                                                                                                                                                                                                                           
               "/dir",                                                                                                                                                                                                                                                        
               "/dir/subdir",                                                                                                                                                                                                                                                 
            +  "/dir/subdir/subsubdir",                                                                                                                                                                                                                                       
            +  "/dir/subdir/subsubdir/file",                                                                                                                                                                                                                                  
               "/dir/subdir2",                                                                                                                                                                                                                                                
               "/dir/subdir2/file",                                                                                                                                                                                                                                           
             }            
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -37,4 +37,4 @@ jobs:
         go-version: stable
     - uses: golangci/golangci-lint-action@3cfe3a4abbb849e10058ce4af15d205b6da42804
       with:
-        version: v1.56.2
+        version: v2.3.1

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -20,8 +20,8 @@ jobs:
         - windows-latest
     runs-on: ${{ matrix.os }}
     steps:
-    - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11
-    - uses: actions/setup-go@0c52d547c9bc32b1aa3301fd7a9cb496313a4491
+    - uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8
+    - uses: actions/setup-go@d35c59abb061a4a6fb18e82ac0862c26744d6ab5
       with:
         go-version: ${{ matrix.go-version }}
     - name: build
@@ -31,10 +31,10 @@ jobs:
   lint:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11
-    - uses: actions/setup-go@0c52d547c9bc32b1aa3301fd7a9cb496313a4491
+    - uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8
+    - uses: actions/setup-go@d35c59abb061a4a6fb18e82ac0862c26744d6ab5
       with:
         go-version: stable
-    - uses: golangci/golangci-lint-action@3cfe3a4abbb849e10058ce4af15d205b6da42804
+    - uses: golangci/golangci-lint-action@4afd733a84b1f43292c63897423277bb7f4313a9
       with:
         version: v2.3.1

--- a/.golangci.bck.yml
+++ b/.golangci.bck.yml
@@ -1,0 +1,127 @@
+linters:
+  enable:
+  - asciicheck
+  - bidichk
+  - bodyclose
+  - containedctx
+  - contextcheck
+  - decorder
+  - dogsled
+  - dupl
+  - dupword
+  - durationcheck
+  - errcheck
+  - errchkjson
+  - errname
+  - errorlint
+  - execinquery
+  - exhaustive
+  - exportloopref
+  - forbidigo
+  - forcetypeassert
+  - gci
+  - ginkgolinter
+  - gocheckcompilerdirectives
+  - gochecksumtype
+  - goconst
+  - gocritic
+  - gocyclo
+  - godot
+  - godox
+  - goerr113
+  - gofmt
+  - gofumpt
+  - goheader
+  - goimports
+  - gomoddirectives
+  - gomodguard
+  - goprintffuncname
+  - gosec
+  - gosimple
+  - gosmopolitan
+  - govet
+  - grouper
+  - importas
+  - inamedparam
+  - ineffassign
+  - ireturn
+  - lll
+  - loggercheck
+  - maintidx
+  - makezero
+  - mirror
+  - misspell
+  - musttag
+  - nakedret
+  - nilerr
+  - nilnil
+  - noctx
+  - nolintlint
+  - nonamedreturns
+  - nosprintfhostport
+  - perfsprint
+  - prealloc
+  - predeclared
+  - promlinter
+  - protogetter
+  - reassign
+  - rowserrcheck
+  - sloglint
+  - spancheck
+  - sqlclosecheck
+  - staticcheck
+  - stylecheck
+  - tagalign
+  - tagliatelle
+  - tenv
+  - testifylint
+  - testpackage
+  - thelper
+  - typecheck
+  - unconvert
+  - unparam
+  - unused
+  - usestdlibvars
+  - wastedassign
+  - whitespace
+  - zerologlint
+  disable:
+  - asasalint
+  - cyclop
+  - depguard
+  - exhaustruct
+  - funlen
+  - gochecknoglobals
+  - gochecknoinits
+  - gocognit
+  - gomnd
+  - interfacebloat
+  - nestif
+  - nlreturn
+  - paralleltest
+  - revive
+  - testableexamples
+  - tparallel
+  - varnamelen
+  - wrapcheck
+  - wsl
+
+linters-settings:
+  gci:
+    sections:
+    - standard
+    - default
+    - prefix(github.com/twpayne/go-vfs)
+  gofumpt:
+    extra-rules: true
+    module-path: github.com/twpayne/go-vfs
+  goimports:
+    local-prefixes: github.com/twpayne/go-vfs
+  misspell:
+    locale: US
+
+issues:
+  exclude-rules:
+  - linters:
+    - goerr113
+    text: "do not define dynamic errors, use wrapped static errors instead"

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -1,129 +1,148 @@
-go-version: '1.21'
-
+version: "2"
 linters:
   enable:
-  - asciicheck
-  - bidichk
-  - bodyclose
-  - containedctx
-  - contextcheck
-  - decorder
-  - dogsled
-  - dupl
-  - dupword
-  - durationcheck
-  - errcheck
-  - errchkjson
-  - errname
-  - errorlint
-  - execinquery
-  - exhaustive
-  - exportloopref
-  - forbidigo
-  - forcetypeassert
-  - gci
-  - ginkgolinter
-  - gocheckcompilerdirectives
-  - gochecksumtype
-  - goconst
-  - gocritic
-  - gocyclo
-  - godot
-  - godox
-  - goerr113
-  - gofmt
-  - gofumpt
-  - goheader
-  - goimports
-  - gomoddirectives
-  - gomodguard
-  - goprintffuncname
-  - gosec
-  - gosimple
-  - gosmopolitan
-  - govet
-  - grouper
-  - importas
-  - inamedparam
-  - ineffassign
-  - ireturn
-  - lll
-  - loggercheck
-  - maintidx
-  - makezero
-  - mirror
-  - misspell
-  - musttag
-  - nakedret
-  - nilerr
-  - nilnil
-  - noctx
-  - nolintlint
-  - nonamedreturns
-  - nosprintfhostport
-  - perfsprint
-  - prealloc
-  - predeclared
-  - promlinter
-  - protogetter
-  - reassign
-  - rowserrcheck
-  - sloglint
-  - spancheck
-  - sqlclosecheck
-  - staticcheck
-  - stylecheck
-  - tagalign
-  - tagliatelle
-  - tenv
-  - testifylint
-  - testpackage
-  - thelper
-  - typecheck
-  - unconvert
-  - unparam
-  - unused
-  - usestdlibvars
-  - wastedassign
-  - whitespace
-  - zerologlint
+    - asciicheck
+    - bidichk
+    - bodyclose
+    - containedctx
+    - contextcheck
+    - decorder
+    - dogsled
+    - dupl
+    - dupword
+    - durationcheck
+    - err113
+    - errchkjson
+    - errname
+    - errorlint
+    - exhaustive
+    - forbidigo
+    - forcetypeassert
+    - ginkgolinter
+    - gocheckcompilerdirectives
+    - gochecksumtype
+    - goconst
+    - gocritic
+    - gocyclo
+    - godot
+    - godox
+    - goheader
+    - gomoddirectives
+    - gomodguard
+    - goprintffuncname
+    - gosec
+    - gosmopolitan
+    - grouper
+    - importas
+    - inamedparam
+    - ireturn
+    - lll
+    - loggercheck
+    - maintidx
+    - makezero
+    - mirror
+    - misspell
+    - musttag
+    - nakedret
+    - nilerr
+    - nilnil
+    - noctx
+    - nolintlint
+    - nonamedreturns
+    - nosprintfhostport
+    - perfsprint
+    - prealloc
+    - predeclared
+    - promlinter
+    - protogetter
+    - reassign
+    - rowserrcheck
+    - sloglint
+    - spancheck
+    - sqlclosecheck
+    - staticcheck
+    - tagalign
+    - tagliatelle
+    - testifylint
+    - testpackage
+    - thelper
+    - unconvert
+    - unparam
+    - usestdlibvars
+    - wastedassign
+    - whitespace
+    - canonicalheader
+    - zerologlint
+    - embeddedstructfieldcheck
+    - exptostd
+    - fatcontext
+    - funcorder
+    - iface
+    - intrange
+    - nilnesserr
+    - usetesting
+    - recvcheck
+    - revive
   disable:
-  - asasalint
-  - cyclop
-  - depguard
-  - exhaustruct
-  - funlen
-  - gochecknoglobals
-  - gochecknoinits
-  - gocognit
-  - gomnd
-  - interfacebloat
-  - nestif
-  - nlreturn
-  - paralleltest
-  - revive
-  - testableexamples
-  - tparallel
-  - varnamelen
-  - wrapcheck
-  - wsl
-
-linters-settings:
-  gci:
-    sections:
-    - standard
-    - default
-    - prefix(github.com/twpayne/go-vfs)
-  gofumpt:
-    extra-rules: true
-    module-path: github.com/twpayne/go-vfs
-  goimports:
-    local-prefixes: github.com/twpayne/go-vfs
-  misspell:
-    locale: US
-
-issues:
-  exclude-rules:
-  - linters:
-    - goerr113
-    text: "do not define dynamic errors, use wrapped static errors instead"
+    - asasalint
+    - cyclop
+    - depguard
+    - exhaustruct
+    - funlen
+    - gochecknoglobals
+    - gochecknoinits
+    - gocognit
+    - interfacebloat
+    - mnd
+    - nestif
+    - nlreturn
+    - paralleltest
+    - revive
+    - testableexamples
+    - tparallel
+    - varnamelen
+    - wrapcheck
+    - wsl
+    - wsl_v5
+  settings:
+    misspell:
+      locale: US
+  exclusions:
+    generated: lax
+    presets:
+      - comments
+      - common-false-positives
+      - legacy
+      - std-error-handling
+    rules:
+      - linters:
+          - err113
+        text: do not define dynamic errors, use wrapped static errors instead
+    paths:
+      - third_party$
+      - builtin$
+      - examples$
+formatters:
+  enable:
+    - gci
+    - gofmt
+    - gofumpt
+    - goimports
+  settings:
+    gci:
+      sections:
+        - standard
+        - default
+        - prefix(github.com/twpayne/go-vfs)
+    gofumpt:
+      module-path: github.com/twpayne/go-vfs
+      extra-rules: true
+    goimports:
+      local-prefixes:
+        - github.com/twpayne/go-vfs
+  exclusions:
+    generated: lax
+    paths:
+      - third_party$
+      - builtin$
+      - examples$

--- a/go.mod
+++ b/go.mod
@@ -1,6 +1,6 @@
 module github.com/twpayne/go-vfs/v5
 
-go 1.21
+go 1.22
 
 require (
 	github.com/alecthomas/assert/v2 v2.6.0

--- a/vfst/fs_test.go
+++ b/vfst/fs_test.go
@@ -3,6 +3,7 @@ package vfst_test
 import (
 	"errors"
 	"io/fs"
+	"os"
 	"path/filepath"
 	"runtime"
 	"testing"
@@ -47,6 +48,7 @@ func TestWalkErrors(t *testing.T) {
 	}
 	for _, tc := range []struct {
 		name          string
+		skip          bool
 		root          any
 		postFunc      func(vfs.FS) error
 		expectedPaths []string
@@ -72,6 +74,7 @@ func TestWalkErrors(t *testing.T) {
 		},
 		{
 			name: "private_subdir",
+			skip: os.Getuid() == 0, // Test fails if run as root, because root can read all files.
 			root: map[string]any{
 				"/dir/subdir/subsubdir/file": "",
 			},
@@ -86,6 +89,7 @@ func TestWalkErrors(t *testing.T) {
 		},
 		{
 			name: "private_subdir_keep_going",
+			skip: os.Getuid() == 0, // Test fails if run as root, because root can read all files.
 			root: map[string]any{
 				"/dir/subdir/subsubdir/file": "",
 				"/dir/subdir2/file":          "",
@@ -103,6 +107,9 @@ func TestWalkErrors(t *testing.T) {
 		},
 	} {
 		t.Run(tc.name, func(t *testing.T) {
+			if tc.skip {
+				t.Skip()
+			}
 			fileSystem, cleanup, err := vfst.NewTestFS(tc.root)
 			assert.NoError(t, err)
 			if tc.postFunc != nil {

--- a/vfst/test.go
+++ b/vfst/test.go
@@ -12,7 +12,7 @@ import (
 )
 
 func init() {
-	umask = fs.FileMode(syscall.Umask(0))
+	umask = fs.FileMode(syscall.Umask(0)) //nolint:gosec
 	syscall.Umask(int(umask))
 }
 
@@ -33,7 +33,7 @@ func TestSysNlink(wantNlink int) PathTest {
 			t.Errorf("fileSystem.Lstat(%q) == %+v, %v, want !<nil>, <nil>", path, info, err)
 			return
 		}
-		if stat, ok := info.Sys().(*syscall.Stat_t); ok && int(stat.Nlink) != wantNlink {
+		if stat, ok := info.Sys().(*syscall.Stat_t); ok && int(stat.Nlink) != wantNlink { //nolint:gosec
 			t.Errorf("fileSystem.Lstat(%q).Sys().(*syscall.Stat_t).Nlink == %d, want %d", path, stat.Nlink, wantNlink)
 		}
 	}

--- a/vfst/testfs.go
+++ b/vfst/testfs.go
@@ -11,6 +11,7 @@ import (
 // A TestFS is a virtual filesystem based in a temporary directory.
 type TestFS struct {
 	vfs.PathFS
+
 	tempDir string
 	keep    bool
 }

--- a/vfst/vfst.go
+++ b/vfst/vfst.go
@@ -77,81 +77,6 @@ func NewBuilder(options ...BuilderOption) *Builder {
 	return b
 }
 
-// build is a recursive helper for Build.
-func (b *Builder) build(fileSystem vfs.FS, path string, i any) error {
-	switch i := i.(type) {
-	case []any:
-		for _, element := range i {
-			if err := b.build(fileSystem, path, element); err != nil {
-				return err
-			}
-		}
-		return nil
-	case *Dir:
-		if parentDir := filepath.Dir(path); parentDir != "." {
-			if err := b.MkdirAll(fileSystem, parentDir, 0o777); err != nil {
-				return err
-			}
-		}
-		if err := b.Mkdir(fileSystem, path, i.Perm); err != nil {
-			return err
-		}
-		entryNames := make([]string, 0, len(i.Entries))
-		for entryName := range i.Entries {
-			entryNames = append(entryNames, entryName)
-		}
-		sort.Strings(entryNames)
-		for _, entryName := range entryNames {
-			if err := b.build(fileSystem, filepath.Join(path, entryName), i.Entries[entryName]); err != nil {
-				return err
-			}
-		}
-		return nil
-	case map[string]any:
-		if err := b.MkdirAll(fileSystem, path, 0o777); err != nil {
-			return err
-		}
-		entryNames := make([]string, 0, len(i))
-		for entryName := range i {
-			entryNames = append(entryNames, entryName)
-		}
-		sort.Strings(entryNames)
-		for _, entryName := range entryNames {
-			if err := b.build(fileSystem, filepath.Join(path, entryName), i[entryName]); err != nil {
-				return err
-			}
-		}
-		return nil
-	case map[string]string:
-		if err := b.MkdirAll(fileSystem, path, 0o777); err != nil {
-			return err
-		}
-		entryNames := make([]string, 0, len(i))
-		for entryName := range i {
-			entryNames = append(entryNames, entryName)
-		}
-		sort.Strings(entryNames)
-		for _, entryName := range entryNames {
-			if err := b.WriteFile(fileSystem, filepath.Join(path, entryName), []byte(i[entryName]), 0o666); err != nil {
-				return err
-			}
-		}
-		return nil
-	case *File:
-		return b.WriteFile(fileSystem, path, i.Contents, i.Perm)
-	case string:
-		return b.WriteFile(fileSystem, path, []byte(i), 0o666)
-	case []byte:
-		return b.WriteFile(fileSystem, path, i, 0o666)
-	case *Symlink:
-		return b.Symlink(fileSystem, i.Target, path)
-	case nil:
-		return nil
-	default:
-		return fmt.Errorf("%s: unsupported type %T", path, i)
-	}
-}
-
 // Build populates fileSystem from root.
 func (b *Builder) Build(fileSystem vfs.FS, root any) error {
 	return b.build(fileSystem, "/", root)
@@ -454,5 +379,80 @@ func TestMinSize(wantMinSize int64) PathTest {
 		if gotSize := info.Size(); gotSize < wantMinSize {
 			t.Errorf("fileSystem.Lstat(%q).Size() == %d, want >=%d", path, gotSize, wantMinSize)
 		}
+	}
+}
+
+// build is a recursive helper for Build.
+func (b *Builder) build(fileSystem vfs.FS, path string, i any) error {
+	switch i := i.(type) {
+	case []any:
+		for _, element := range i {
+			if err := b.build(fileSystem, path, element); err != nil {
+				return err
+			}
+		}
+		return nil
+	case *Dir:
+		if parentDir := filepath.Dir(path); parentDir != "." {
+			if err := b.MkdirAll(fileSystem, parentDir, 0o777); err != nil {
+				return err
+			}
+		}
+		if err := b.Mkdir(fileSystem, path, i.Perm); err != nil {
+			return err
+		}
+		entryNames := make([]string, 0, len(i.Entries))
+		for entryName := range i.Entries {
+			entryNames = append(entryNames, entryName)
+		}
+		sort.Strings(entryNames)
+		for _, entryName := range entryNames {
+			if err := b.build(fileSystem, filepath.Join(path, entryName), i.Entries[entryName]); err != nil {
+				return err
+			}
+		}
+		return nil
+	case map[string]any:
+		if err := b.MkdirAll(fileSystem, path, 0o777); err != nil {
+			return err
+		}
+		entryNames := make([]string, 0, len(i))
+		for entryName := range i {
+			entryNames = append(entryNames, entryName)
+		}
+		sort.Strings(entryNames)
+		for _, entryName := range entryNames {
+			if err := b.build(fileSystem, filepath.Join(path, entryName), i[entryName]); err != nil {
+				return err
+			}
+		}
+		return nil
+	case map[string]string:
+		if err := b.MkdirAll(fileSystem, path, 0o777); err != nil {
+			return err
+		}
+		entryNames := make([]string, 0, len(i))
+		for entryName := range i {
+			entryNames = append(entryNames, entryName)
+		}
+		sort.Strings(entryNames)
+		for _, entryName := range entryNames {
+			if err := b.WriteFile(fileSystem, filepath.Join(path, entryName), []byte(i[entryName]), 0o666); err != nil {
+				return err
+			}
+		}
+		return nil
+	case *File:
+		return b.WriteFile(fileSystem, path, i.Contents, i.Perm)
+	case string:
+		return b.WriteFile(fileSystem, path, []byte(i), 0o666)
+	case []byte:
+		return b.WriteFile(fileSystem, path, i, 0o666)
+	case *Symlink:
+		return b.Symlink(fileSystem, i.Target, path)
+	case nil:
+		return nil
+	default:
+		return fmt.Errorf("%s: unsupported type %T", path, i)
 	}
 }

--- a/walk.go
+++ b/walk.go
@@ -11,7 +11,7 @@ import (
 )
 
 // SkipDir is fs.SkipDir.
-var SkipDir = fs.SkipDir
+var SkipDir = fs.SkipDir //nolint:errname
 
 // A LstatReadDirer implements all the functionality needed by Walk.
 type LstatReadDirer interface {

